### ISSUE TITLE
Processing very large pathological images with pyvips

### ISSUE THREAD

**Issue by viktoriaas** [libvips/libvips#1254]
Hi @jcupitt ,

as a part of my work, I need to convert images from .czi file format to VIPS using Python 3. 
If you are not familiar with this format, I'll shortly describe it. CZI file format is a format produced by Carl Zeiss microscopes, it can store multiple planes, slices or even time dimensions. The data in this format are saved in so-called Subblocks, in numpy arrays. You can imagine these subblocks as tiles that comprise whole image.

Right now I'm able to somehow obtain data from these subblocks as a numpy array and via  pyvips.Image.new_from_memory make a VIPS image from every tile. I have these information about every tile: exact x and y position in the big image and width and height of tile.

What I need to do is join all of the tiles into one big VIPS image. However, there are several problems, mainly memory issues. The CZI images are often quite big (over 40 000 x 40 000 pixels) and in one czi image, I can have f.i. 5 planes which means that I have to make 5 images with resolution 40 000 x 40 000. On top of this, I need to read the data from numpy array which costs me some extra memory.

I have already tried several approaches of solving this:
1. I found this issue : https://github.com/libvips/libvips/issues/1135
First I tried using what the person who asked tried - making black image, rawsaving,rawloading and inserting, then saving to vips. This took too much memory and time.
2. Then I tried almost same version but instead of inserting, I used draw_image operation. This works better, it is able to write smaller images quite nicely but for slightly bigger images, needs at least 16GB of memory. And images can get even bigger so I can imagine that memory usage will go up.
3. So then I started to think about using C-library function im_writeline which should be able to write lines of pixels really quickly and without excessive memory usage. While browsing ideas, I found this yesterday: https://stackoverflow.com/questions/47852390/making-a-huge-image-mosaic-with-pyvips
This is the approach I'm very likely looking for. I have already tried using the pyvips.Image.arrayjoin operation in my code and it works quite nice, I guess that it is faster (can't tell exactly beacuse on my home laptop I can process only smaller images, difference in performance is better visible on bigger images). But there is one big issue - the small tiles should overlap.
With using draw_image I was able to draw image on exact position so the leftmost and rightmost parts of images overlapped and nicely made up whole image. However, with arrayjoin I am unable to come up with way how to compute the "shifting" or how to pass this function fact, that the tiles should't be next to each other but slightly overlap (not only in left-right direction but also in top-bottom)

What I am able to find out is: number of tiles in one line, number of lines, width and height of the big image I should have and the information I mentioned about tiles.

The crucial part of the code is:

    `im = None
     tiles = []
     for line in linet: #lines in image, saved in dict for reason
         tiles_num = linet[line]
         m = tiles_num
         print('Building row ', line, '/', len(linet))
         for x in range(tiles_num): #tiles in one line
             de = m_subblocks[index].directory_entry
             # operations needed to obtain pixel data from tile
             de = m_subblocks[index].directory_entry
            for entry in de.dimension_entries:
                if entry.dimension == 'X':
                    xpos = entry.start
                if entry.dimension == 'Y':
                    ypos = entry.start
            subdata = np.squeeze(m_subblocks[index].data())
            dir_shape = subdata.shape
            ydim, xdim, channels = dir_shape
            linear = subdata.reshape(ydim * xdim * channels)
            index += 1
             vi = pyvips.Image.new_from_memory(linear.data, xdim, ydim, 3,dtype_to_format[str(linear.dtype)])
            tiles.append(vi)`

    `im = pyvips.Image.arrayjoin(tiles, across=m,shim=12)

    print("Saving created VIPS image to " + location)
    im.write_to_file(location)`

once I added "shim=12" param to arrayjoin, just to look what it makes.It nicely showed how the tiles are next to each other and it can be observed where they should overlap. 
Image on the left is what I want to have (this was made using draw_image). Center image shows how arrayjoin creates image and right image is with shim=12, you can see on the tiles that there are same "pink areas" that should obviously overlap. 

![img](https://user-images.githubusercontent.com/44999986/54169502-399c8e80-4473-11e9-882e-8b3dcba8784c.png)

So I would be very thankful if you could help me with solving my overlapping problem or if you have any other idea how should I make processing images faster, I would be very grateful.

Thanks in advance, 
Viktoria

**Comment by jcupitt** [comment]
Hello, sorry for the delay, I'll try to answer later today.

**Comment by viktoriaas** [comment]
Hello, 
I surely can wait. In the meantime, I found out that approach with draw_image is not working on images I will be frequently processing. 
I have a czi file with 5 planes == 5 images, each of them with Height: 128523, Width:141328 px.
When I try running the program, I get this error:
pyvips.error.Error: unable to copy to memory
  vips_tracked: out of memory --- size == 51967MB.

I surely have enough space on disk for the big black image and RAM is 64GB. When I tried writing to disk every tile, it worked so the problem is writing them to big image. I counted that big image should me made of 9764 tiles as this much has been written to disc. 
So I guess I really need to use another way of processing.

**Comment by martinweihrauch** [comment]
Hi viktoriaas:

whats your ultimate goal for the czi images?

**Comment by viktoriaas** [comment]
> Hi viktoriaas:
> 
> whats your ultimate goal for the czi images?

**Comment by viktoriaas** [comment]
Hello @martinweihrauch, 

actually, I should close this issue because I rewrote the code into C++. I
n Python, it was incredibly slow beacuse of one line in Python wrapper in file pyvips/voperation.py, function 
`set(self, name, flags, match_image, value)`
line
`value = value.copy().copy_memory()` 

When I used function draw_image, it used this set function to copy parameters, I guess. However, this double copying increases time and memory consumption enormously. I had image with about 9000 tiles. When I used Python code, after *13 hours* it converted only 700 tiles with memory usage of 102G. getting image from memory was fast, I would say instatntly, but after calling draw_image, the code paused for every tile for 50sec - 70seconds (because of copying).
In C++, same image, converting of whole image took 43minutes. 
I have already deleted the Python code, it was kinda useless
